src/domain-services/flows/strategy/impl/get-flowIds-flow-category-conditions-strategy-impl.ts

```diff
@@ -62,39 +63,91 @@ export class GetFlowIdsFromCategoryConditionsStrategyImpl
       }
     }
 
-    // Search categoriesRef with categoriesID IN and categoriesIdsFromShortcutFilterIN
-    // and categoriesIdsFromShortcutFilterNOTIN
+    // Search categoryRef rows for the relevant categories, then post-process
+    // so multiple IN filters are treated with AND semantics. We fetch rows
+    // (not distinct) so we can count which categories each flow has, then
+    // only keep flows that contain all required IN categories and none of the
+    // forbidden NOT_IN categories.
     const where: Condition<InstanceOfModel<Database['categoryRef']>> = {
       objectType: 'flow',
     };
 
-    if (categoriesIdsFromShortcutFilterNOTIN.length > 0) {
-      where['categoryID'] = {
-        [Op.NOT_IN]: categoriesIdsFromShortcutFilterNOTIN,
-      };
-    }
-
     const categoriesIDsIN = [
       ...categoriesIds,
       ...categoriesIdsFromShortcutFilterIN,
     ];
+    const categoriesIDsNOTIN = categoriesIdsFromShortcutFilterNOTIN;
+
+    // To limit the rows fetched, request refs whose categoryID is in the
+    // union of required IN and forbidden NOT_IN (if any). If there are no
+    // constraints, the where will only include objectType and return all refs
+    const fetchCategoryIDs = [
+      ...new Set([...categoriesIDsIN, ...categoriesIDsNOTIN]),
+    ];
+
+    if (fetchCategoryIDs.length > 0) {
+      // Use the raw values from DB for the IN filter. We'll stringify for
+      // comparisons when grouping below.
+      where['categoryID'] = { [Op.IN]: fetchCategoryIDs };
+    }
+
+    // Fetch all matching categoryRef rows so we can
+    // determine which categories each flow has.
+    const categoriesRef = await models.categoryRef.find({ where });
+
+    // Group refs by flow (objectID + versionID) and collect the set of
+    // categoryIDs attached to each flow.
+    const flowMap = new Map<
+      string,
+      { id: FlowId; versionID: number; categorySet: Set<string> }
+    >();
 
-    if (categoriesIDsIN.length > 0) {
-      where['categoryID'] = { [Op.IN]: categoriesIDsIN };
+    for (const catRef of categoriesRef) {
+      const key = `${catRef.objectID}::${catRef.versionID}`;
+      let entry = flowMap.get(key);
+      if (!entry) {
+        entry = {
+          id: createBrandedValue(catRef.objectID),
+          versionID: catRef.versionID,
+          categorySet: new Set<string>(),
+        };
+        flowMap.set(key, entry);
+      }
+      entry.categorySet.add(String(catRef.categoryID));
```

Comment:
Nitpick: `getOrCreate` is an useful utility function for this case.

```suggestion
      const key = `${catRef.objectID}::${catRef.versionID}`;

      const entry = getOrCreate(
        flowMap,
        key,
        () => ({
          id: createBrandedValue(catRef.objectID),
          versionID: catRef.versionID,
          categorySet: new Set<string>(),
        })
      );

      entry.categorySet.add(String(catRef.categoryID));
```
or
```suggestion
      getOrCreate(flowMap, `${catRef.objectID}::${catRef.versionID}`, () => ({
        id: createBrandedValue(catRef.objectID),
        versionID: catRef.versionID,
        categorySet: new Set<string>(),
      })).categorySet.add(String(catRef.categoryID));
```